torch/nn/parallel/distributed.py

```diff
@@ -117,14 +122,34 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
         else:
             self._module_copies = [self.module]
 
+        # Split parameters into type buckets so that parameter sync (broadcast)
+        # can operates on mixed parameter types. (e.g. mixed half and float)
+        self.param_type_buckets = {}
+        for device_idx, module in enumerate(self._module_copies):
+            for p in module.parameters():
+                tp = type(p.data)
+                if tp == torch.cuda.HalfTensor and \
+                        dist._backend != dist.dist_backend.NCCL:
+                    raise RuntimeError("DistributedDataParallel currently only "
+                                       "supports half precision parameters "
+                                       "with NCCL backend")
+                if tp not in self.param_type_buckets:
+                    self.param_type_buckets[tp] = \
+                        [[] for _ in range(len(self.device_ids))]
```

Comment:
make `param_type_buckets` a default dict with a lambda that creates this list as a constructor?